File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
private:
    vector<int> toposort(vector<vector<int>> &adj, int numCourses){
        queue<int> q;
        vector<int> ans;
        vector<int> indegree(numCourses,0);
        for (auto nodes : adj) {
            for (auto node : nodes) {
                indegree[node]++;
            }
        }
        for(int i=0;i<numCourses;i++){
            if(indegree[i]==0) q.push(i);
        }

        while(!q.empty()){
            int node=q.front();
            ans.push_back(node);
            q.pop();
            for(auto it:adj[node]){
                indegree[it]--;
            if(indegree[it]==0){
                q.push(it);
            }
            }
        }
        return ans;
    }
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        for(auto edge:prerequisites){
            adj[edge[1]].push_back(edge[0]);
        }

        //now detect cycle
        vector<int> topo=toposort(adj,numCourses);
        if(topo.size()<numCourses){
            return false;
        }
        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
from collections import defaultdict
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = defaultdict(list)

        for course,preq in prerequisites:
            adj[preq].append(course)

        visited = set()
        path = set()
        

        def dfs(node):
            if node in path:
                return True
            
            if node in visited:
                return False


            visited.add(node)
            path.add(node)

            for nei in adj[node]:
                if dfs(nei):
                    return True
            
            path.remove(node)
            return False

        for course in range(numCourses):
            if course not in visited:
                if dfs(course):
                    return False
        

        return True